Add contribution guidelines

imphil opened the pull request:
To revive the cocotb development community a group of developers has
agreed to create a contribution guide which formalizes existing
processes and adds new one.

The goal of this guide is to eliminate bottlenecks in the development
process, and ultimately provide a pleasant development experience for
all parties involved.

The most critical aspect for developers is a fast process from opening a
PR to merging changes. This guide tries to streamline reviews by
- stating explicitly that reviews are expected within one week,
- requiring only a single positive review in cases where no maintainer
  objects,
- making "merging" a process step which can be done by any maintainer
  if a well-defined process (including reviews) has been followed.

The guide also helps to relieve stress from maintainers by sharing the
review load among multiple parties.

Further parts of this guide are more on the logistics side, including a
description of GitHub labels.

The release process isn't very well defined at the moment, I see this as
a first shot and expect clearifications in the future.

The list of maintainers is based on the notes I've taken during the
conference call, please correct and extend as needed. Also please add
areas of expertise if possible.

Finally, the code of conduct is something we hopefully never need to
make use of. It's standard practice these days to include one; I've
chosen the FOSSi Foundation one, which is itself based on the Apache
Code of Conduct.

imphil commented:
This is the first draft of the contribution guidelines as discussed in the call yesterday. I invite everyone to comment extensively and/or to agree/disagree with it. I'd love to get a wide consensus on this one, and I'm prepared to make significant changes to the document if needed.

Some of the terminology and the ideas are borrowed from the ZeroMQ C4 (https://rfc.zeromq.org/spec:42/C4/). I didn't copy all concepts to make the document easier to read, and to make it more applicable to cocotb.

cmarqu commented on CONTRIBUTING.md:
Typo, should be "specifically".

imphil replied:
fixed, thanks.

cmarqu commented on CONTRIBUTING.md:
We don't have IRC (right?), but we could use Gitter.

imphil replied:
good catch, I've added the Gitter link.

cmarqu commented on CONTRIBUTING.md:
You appear to want to sort by last name, in which case, Julius should move one position up.

imphil replied:
good catch, fixed as well.

imphil commented on CONTRIBUTING.md:
fixed.

imphil commented:
Thanks for the review @cmarqu, I've pushed an  updated version.

imphil commented:
@stuarthodgson I've modified the text a bit to incorporate your feedback.

I also added a sentence that reviews should be done by somebody else than the author of the code (only applicable in case the author is also a maintainer).
